### DISCUSSION TITLE
Does SkiaSharp work on Linux? How?

### DISCUSSION THREAD

**Discussion by bmitc** [mono/SkiaSharp#2538] [Q&A]
What are the steps required to use SkiaSharp on Linux on major distributions like Ubuntu or Linux Mint?

See also:
* https://github.com/mono/SkiaSharp/issues/2350
* https://stackoverflow.com/questions/74882243/how-do-i-properly-build-a-net-7-application-using-skiasharp-on-linux-getting-a

I know about the existence of:
* https://www.nuget.org/packages/SkiaSharp.NativeAssets.Linux
* https://www.nuget.org/packages/SkiaSharp.NativeAssets.Linux.NoDependencies

but I don't know how to use them, and they're undocumented as far as I can tell.

Specifically, I have tried using SkiaSharp on Ubuntu and Linux Mint with these, and I cannot get `GRGlInterface.Create` to not throw an exception. So I am either doing something with regards to SkiaSharp on Linux or there is a bug. This discussion is trying to address the former.

**Comment by mattleibow** [comment]
Using SkiaSharp with the GPU requires that you first set up the GPU context. In the case of OpenGL, you need a GL context that is current, and then the GLInterface will be constructed using that. If you have mutilple GL contexts then you will also need multiple GRContexts.

**Reply by bmitc** [reply]
Hi @mattleibow, thank you for the response! However, I'm not sure you're explaining any new information not already covered in the linked issue and StackOverflow post. Did you happen to take a look at them?

To be specific, what is wrong with the following code? You only need the `Silk.NET.GLFW` and `SkiaSharp` NuGet packages to run it. On Linux Mint or Ubuntu, the window opens, and then the application immediately crashes, apparently while creating the `GRGlInterface` given that only a few proc names are printed out. Exit code 139 is a segmentation fault. This code works perfectly fine on Windows. I don't currently have a macOS machine to test on.

```fsharp
open FSharp.NativeInterop

open Silk.NET.GLFW
open SkiaSharp

#nowarn "9"

let width, height = 800, 600

let glfw = Glfw.GetApi()
glfw.Init() |> printfn "Initialized?: %A"

let window = glfw.CreateWindow(width, height, "Test Window", NativePtr.ofNativeInt 0n, NativePtr.ofNativeInt 0n)
printfn "Window: %A" window
glfw.MakeContextCurrent(window)
let mutable error = nativeint<byte> 1uy |> NativePtr.ofNativeInt
glfw.GetError(&error) |> printfn "Error: %A"

let grGlInterface = GRGlInterface.Create(fun name ->
    printfn "Process/procedure name: %A" name
    glfw.GetProcAddress name)

if not(grGlInterface.Validate()) then
    raise (System.Exception("Invalid GRGlInterface"))

let grContext = GRContext.CreateGl(grGlInterface)
let grGlFramebufferInfo = new GRGlFramebufferInfo(0u, SKColorType.Rgba8888.ToGlSizedFormat()) // 0x8058
let grBackendRenderTarget = new GRBackendRenderTarget(width, height, 1, 0, grGlFramebufferInfo)
let surface = SKSurface.Create(grContext, grBackendRenderTarget, GRSurfaceOrigin.BottomLeft, SKColorType.Rgba8888)
let canvas = surface.Canvas
grContext.ResetContext()

while not (glfw.WindowShouldClose(window)) do
    glfw.PollEvents()
    canvas.Clear(SKColors.WhiteSmoke)
    canvas.Flush()
    glfw.SwapBuffers(window)
        
glfw.DestroyWindow(window)
glfw.Terminate()
```

As already mentioned in the links, this returns:

```bash
$ dotnet run --project SilkNETPerformanceTest
Initialized?: true
Window: 93998088633456n
Error: NoError
Process/procedure name: "glGetString"
Process/procedure name: "glGetString"
Process/procedure name: "glGetStringi"
Process/procedure name: "glGetIntegerv"
Process/procedure name: "eglQueryString"
Process/procedure name: "eglGetCurrentDisplay"

$ echo $?
139

$ glxinfo | grep "OpenGL version"
OpenGL version string: 4.6 (Compatibility Profile) Mesa 22.0.5
```

**Comment by ikasoba** [comment]
In my environment, inserting `glfw.WindowHint(WindowHintContextApi.ContextCreationApi, ContextApi.EglContextApi)` after `glfw.Init()` works.

The following information was helpful in addressing this issue.

https://github.com/rust-skia/rust-skia/issues/511
